src/devices/src/virtio/balloon/device.rs

```diff
@@ -256,58 +256,86 @@ impl Balloon {
         METRICS.balloon.inflate_count.inc();
 
         let queue = &mut self.queues[INFLATE_INDEX];
-        let mut pages = Vec::with_capacity(MAX_PAGES_IN_DESC);
+        // A buffer used as pfn accumulator during descriptor processing.
+        // It is important that it does not grow by overrunning it's predefined capacity, in
+        // which case it will issue a mmap syscall which will terminate the process via seccomp
+        // violation.
+        let mut pages = Vec::with_capacity(MAX_PAGE_COMPACT_BUFFER);
         let mut needs_interrupt = false;
 
-        while let Some(head) = queue.pop(&mem) {
-            let len = head.len;
-            if !head.is_write_only() && len % SIZE_OF_U32 as u32 == 0 {
-                for index in (0..len).step_by(SIZE_OF_U32) {
-                    let addr = head
-                        .addr
-                        .checked_add(index as u64)
-                        .ok_or(BalloonError::MalformedDescriptor)?;
-
-                    let page_frame_number = mem
-                        .read_obj::<u32>(addr)
-                        .map_err(|_| BalloonError::MalformedDescriptor)?;
-
-                    pages.push(page_frame_number);
+        // Loop until we consume the entire queue.
+        while !queue.is_empty(&mem) {
+            // This loop processes descriptors and acummulates the pfns in pages buffer.
+            // Breaks out when there is not enough space to completely process a desc.
+            while let Some(head) = queue.pop(&mem) {
+                let mut len = head.len as usize;
+                let max_len = MAX_PAGES_IN_DESC * SIZE_OF_U32;
+
+                if !head.is_write_only() && len % SIZE_OF_U32 == 0 {
+                    // Break loop if page buffer will be overrun by adding pages from current desc.
+                    if pages.capacity() - pages.len() < len as usize / SIZE_OF_U32 {
```

Comment:
`pages.capacity()` can change on a `push` or other conditions, let's check against what we care about:

```suggestion
                    if MAX_PAGE_COMPACT_BUFFER - pages.len() < len as usize / SIZE_OF_U32 {
```